subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/FileChange.java

```diff
@@ -18,22 +18,55 @@
 
 import com.google.common.base.Objects;
 import org.gradle.api.tasks.incremental.InputFileDetails;
+import org.gradle.internal.nativeintegration.filesystem.FileType;
 
 import java.io.File;
 
 public class FileChange implements TaskStateChange, InputFileDetails {
     private final String path;
     private final ChangeType change;
     private final String fileType;
+    private final FileType previousType;
+    private final FileType currentType;
```

Comment:
- [ ] All these `.*Type` fields are confusing. I think we should at least rename `fileType` to something more descriptive. We could also call the new fields `previousFileType` and `currentFileType`.